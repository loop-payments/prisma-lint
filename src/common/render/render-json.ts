
import { keyViolationListPairs } from '#src/common/render/render-util.js';
import type { Violation } from '#src/common/violation.js';

export const renderViolationsJson = (violations: Violation[]) => {
  const jsonObject = renderViolationsJsonObject(violations);
  return JSON.stringify(jsonObject);
};

export const renderViolationsJsonObject = (
  violations: Violation[],
): Record<string, any>[] => {
  const pairs = keyViolationListPairs(violations);
  return pairs.flatMap(([_, violations]) => {
    const first = violations[0];
    const { fileName } = first;
    const rawLocation = first.field?.location ?? first.model.location;
    if (!rawLocation) {
      throw new Error('No location');
    }
    const {
      startLine,
      startColumn,
      startOffset,
      endLine,
      endColumn,
      endOffset,
    } = rawLocation;
    const location = {
      startLine,
      startColumn,
      startOffset,
      endLine,
      endColumn,
      endOffset,
    };
    return violations.map((v) => ({
      ruleName: v.ruleName,
      message: v.message,
      fileName,
      location,
    }));
  });
};
