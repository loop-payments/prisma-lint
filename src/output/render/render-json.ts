import type { Violation } from '#src/common/violation.js';
import { keyViolationListPairs } from '#src/output/render/render-util.js';

export const renderViolationsJson = (violations: Violation[]) => {
  const jsonObject = renderViolationsJsonObject(violations);
  return JSON.stringify(jsonObject);
};

export const renderViolationsJsonObject = (
  violations: Violation[],
): Record<string, any>[] => {
  const pairs = keyViolationListPairs(violations);
  return pairs.flatMap(([_, violations]) => {
    const first = violations[0];
    const { fileName } = first;
    const rawLocation =
      first.field?.location ?? first.model?.location ?? first.enum?.location;
    if (!rawLocation) {
      throw new Error('No location');
    }
    const { startLine, startColumn, endLine, endColumn } = rawLocation;
    if (
      startLine === undefined ||
      startColumn === undefined ||
      endLine === undefined ||
      endColumn === undefined
    ) {
      throw new Error('No location');
    }
    const location = {
      startLine,
      startColumn,
      endLine,
      endColumn,
    };
    return violations.map((v) => ({
      ruleName: v.ruleName,
      message: v.message,
      fileName,
      location,
    }));
  });
};
