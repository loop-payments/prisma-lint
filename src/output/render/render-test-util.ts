import type { Model } from '@mrleebo/prisma-ast';

import type { Violation } from '#src/common/violation.js';

export const MOCK_SOURCE_CODE = `model Foo {
  id     Int    @id
  displayName String
}`;

const model: Model = {
  type: 'model',
  name: 'Foo',
  properties: [],
  location: {
    startLine: 1,
    startColumn: 1,
    startOffset: 1,
    endLine: 1,
    endColumn: 9,
    endOffset: 1,
  },
};
const fileName = 'test.prisma';
export const MOCK_VIOLATIONS: Violation[] = [
  {
    ruleName: 'fake-model-rule-name',
    message: 'Fake model rule violation message',
    fileName,
    model,
  },
  {
    ruleName: 'fake-field-rule-name',
    message: 'Fake field rule violation message',
    fileName,
    model,
    field: {
      type: 'field',
      name: 'displayName',
      fieldType: 'String',
      location: {
        startLine: 2,
        startColumn: 3,
        startOffset: 1,
        endLine: 2,
        endColumn: 12,
        endOffset: 1,
      },
    },
  },
];
