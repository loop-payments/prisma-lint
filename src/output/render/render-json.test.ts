import { renderViolationsJsonObject } from '#src/output/render/render-json.js';
import { MOCK_VIOLATIONS } from '#src/output/render/render-test-util.js';

describe('render json', () => {
  it('matches snapshot', () => {
    const result = renderViolationsJsonObject(MOCK_VIOLATIONS);
    expect(result).toBeDefined();
    expect(result).toMatchSnapshot();
  });
});
