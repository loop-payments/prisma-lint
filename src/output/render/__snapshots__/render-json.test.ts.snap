// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render json matches snapshot 1`] = `
[
  {
    "fileName": "test.prisma",
    "location": {
      "endColumn": 9,
      "endLine": 1,
      "startColumn": 1,
      "startLine": 1,
    },
    "message": "Fake model rule violation message",
    "ruleName": "fake-model-rule-name",
  },
  {
    "fileName": "test.prisma",
    "location": {
      "endColumn": 12,
      "endLine": 2,
      "startColumn": 3,
      "startLine": 2,
    },
    "message": "Fake field rule violation message",
    "ruleName": "fake-field-rule-name",
  },
]
`;
